field.go

```diff
@@ -61,6 +74,34 @@ func (f *field) Type() FieldType                              { return f.typ }
 func (f *field) setMessage(m Message)                         { f.msg = m }
 func (f *field) setOneOf(o OneOf)                             { f.oneof = o }
 
+func (f *field) InRealOneOf() bool {
+	return f.InOneOf() && !f.desc.GetProto3Optional()
+}
+
+func (f *field) HasPresence() bool {
+	if f.InOneOf() {
+		return true
+	}
+
+	if f.Type().IsEmbed() {
+		return true
+	}
+
+	if f.HasOptionalKeyword() ||
+		(f.Syntax() == Proto2 && !f.Type().IsRepeated() && !f.Type().IsMap()) {
```

Comment:
nit I would rewrite this as:

```
if (!f.Type().IsRepeated() && !f.Type().IsMap()) {
	if f.Syntax() == Proto2 { 
		return true
	}
    return f.HasOptionalKeyword()
}
```

No deep reason, it just made more sense to me. LMK if you disagree.